lib/ecto/type.ex

```diff
@@ -367,6 +371,14 @@ defmodule Ecto.Type do
     {:ok, Application.get_env(:ecto, :json_library).decode!(value)}
   end
 
+  def load({:map, model}, value) when is_binary(value) do
+    load({:map, model}, Application.get_env(:ecto, :json_library).decode!(value))
+  end
+
+  def load({:map, model}, map) when is_map(map) do
+    {:ok, model.__schema__(:load, "", map, %{})}
```

Comment:
Instead of passing an empty string, let's pass nil, and change the code to do `source || __schema__(:source)`.